tokio-trace/tokio-trace-core/src/callsite.rs

```diff
@@ -24,6 +24,27 @@ struct Registry {
     dispatchers: Vec<dispatcher::Registrar>,
 }
 
+impl Registry {
+    fn establish_interest(&self, callsite: &'static Callsite) {
+        let meta = callsite.metadata();
+
+        self.dispatchers.iter().for_each(|registrar| {
+            if let Some(interest) = registrar.try_register(meta) {
+                callsite.add_interest(interest)
+            }
+        });
+    }
+
+    fn reestablish(&mut self) {
+        self.dispatchers.retain(Registrar::is_alive);
+
+        self.callsites.iter().for_each(|&callsite| {
+            callsite.clear_interest();
+            self.establish_interest(callsite);
```

Comment:
I’m just skimming through at the moment but this feels racy. Could this cause unrelated subscribers to lose events?

Reply:
@carllerche I believe this is always guarded by the global mutex around the registry?

Reply:
The registry is guarded, but the call site is not. This function clears interest and then establishes interest. If the call site receives an event between these two steps, the event will be lost even if a subscriber is interested in it.

Reply:
@carllerche ah, yeah, I see it now. That definitely could be an issue (and might explain why @yaahallo saw some apparently non-deterministic CI failures on this branch?).

The only solution I can come up with is to change the `Callsite` API so that instead of having a `clear_interest` method, it has a `set_interest` that's expected to atomically replace the previous `Interest`. In that case, the registry would be responsible for aggregating the interests returned by each subscriber and then providing that result to the callsite. This way, the callsite never enters a temporary state where it has no interest.

I think it's _technically_ possible to do the above without a breaking change, by deprecating `clear_interest`, adding a new method `set_interest` with a default impl. However, I think the default impl would have to be something like
```rust
self.clear_interest();
self.add_interest(new_interest);
```
which still has the potentially racy behaviour. I'm not a big fan of having an incorrect default impl, so it might be better to just wait for the next breaking release to add callsite cache invalidation?

It's possible I'm missing something and there's another solution without breaking existing APIs...

Reply:
would it be possible to not deprecate clear_interest and leave the existing API alone but add set_interest for usage in the new `invalidate_interest` or w/e it ends up being called function, that way the new function isn't racy but presumably the old behaviour is unchanged..

Reply:
We might as well issue a breaking release to fix the issues.